Reland `optimized-compiler-builtins` config

jyn514 opened the pull request:
this is possible now that compiler-builtins contains the fix in https://github.com/rust-lang/compiler-builtins/pull/532.

in particular, this makes the `c` feature for compiler-builtins an explicit opt-in, rather than silently detected by whether `llvm-project` is checked out on disk.
exposing this is necessary because the `cc` crate doesn't support cross-compiling to MSVC, and we want people to be able to run `x check --target foo` regardless of whether they have a c toolchain available.

this also uses the new option in CI, where we *do* want to optimize compiler_builtins.

the new option is off by default for the `dev` channel and on otherwise.

Fixes https://github.com/rust-lang/rust/issues/102560. Fixes https://github.com/rust-lang/rust/issues/101172. Helps with https://github.com/rust-lang/rust/issues/105065 (although there's some weirdness there - it's still broken when `optimized-compiler-builtins` is set to `true`).

r? @tmandry cc @Mark-Simulacrum

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3168746373/jobs/5160164384) [(plain)](https://github.com/rust-lang/rust/commit/3e238a318e91e9f68ad2794de36b890879149fc4/checks/8663401892/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling toml v0.5.9
error[E0424]: expected value, found module `self`
    --> dist.rs:1860:37
     |
1846 | fn maybe_install_llvm(builder: &Builder<'_>, target: TargetSelection, dst_libdir: &Path) -> bool {
     |    ------------------ this function can't have a `self` parameter
...
1860 |     let should_install_llvm = match self.config.target_config.get(&target) {
     |                                     ^^^^ `self` value is a keyword only available in methods with a `self` parameter
error[E0422]: cannot find struct, variant or union type `Target` in this scope
    --> dist.rs:1863:14
     |
     |
1863 |         Some(Target { llvm_config: Some(_), .. }) => self.config.llvm_from_ci && target == self.config.build,
     |
help: consider importing one of these items
     |
11   | use crate::DependencyType::Target;
11   | use crate::DependencyType::Target;
     |
11   | use crate::Target;
     |

error[E0424]: expected value, found module `self`
    --> dist.rs:1863:54
     |
1846 | fn maybe_install_llvm(builder: &Builder<'_>, target: TargetSelection, dst_libdir: &Path) -> bool {
     |    ------------------ this function can't have a `self` parameter
...
1863 |         Some(Target { llvm_config: Some(_), .. }) => self.config.llvm_from_ci && target == self.config.build,
     |                                                      ^^^^ `self` value is a keyword only available in methods with a `self` parameter
error[E0424]: expected value, found module `self`
    --> dist.rs:1863:92
     |
     |
1846 | fn maybe_install_llvm(builder: &Builder<'_>, target: TargetSelection, dst_libdir: &Path) -> bool {
     |    ------------------ this function can't have a `self` parameter
...
1863 |         Some(Target { llvm_config: Some(_), .. }) => self.config.llvm_from_ci && target == self.config.build,
     |                                                                                            ^^^^ `self` value is a keyword only available in methods with a `self` parameter
error[E0422]: cannot find struct, variant or union type `Target` in this scope
    --> dist.rs:1864:14
     |
     |
1864 |         Some(Target { llvm_config: None, .. }) | None => true,
     |
help: consider importing one of these items
     |
11   | use crate::DependencyType::Target;
```

</details>

tmandry commented on src/bootstrap/lib.rs:
```diff
@@ -846,18 +846,19 @@ impl Build {
+    /// This does not necessarily imply that we're managing the `llvm-project` submodule.
     fn is_rust_llvm(&self, target: TargetSelection) -> bool {
         match self.config.target_config.get(&target) {
+            // We're using a pre-built version of LLVM, but the user has promised that pre-built version has our patches.
```
We may not be using a pre-built version of LLVM. Specifically the use case I want to support is cd into `src/llvm-project`, check out a vanilla commit, and set `submodules = false`, `llvm-has-rust-patches = false` in config.

jyn514 replied:
That seems totally bizarre, though. Why would you want that?

tmandry replied:
To build against a more recent version of LLVM.

If you separately build and supply your own llvm-config, then you hit the other assumptions in here about not needing to distribute the LLVM libraries because it's probably system installed. So basically this was unsupported before.

Plus it's desirable to have bootstrap manage the configuration of LLVM if all you want to vary is the version.

tmandry commented on src/bootstrap/compile.rs:
```diff
@@ -313,12 +313,17 @@ pub fn std_cargo(builder: &Builder<'_>, target: TargetSelection, stage: u32, car
     let compiler_builtins_c_feature = if builder.config.optimized_compiler_builtins {
         if !builder.is_rust_llvm(target) {
```
You can use `compiler-rt` without Rust patches. I think the important thing is that `compiler-rt` is built from the same sources as LLVM. So.. either `llvm-config` is unset, or llvm_from_ci is active and we're managing submodules?

tmandry commented:
@rustbot author

bors commented:
:umbrella: The latest upstream changes (presumably #103797) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jyn514 commented:
@tmandry I am still very confused how `llvm_has_rust_patches` is meant to be used. The only place I see it used is in `Build::is_rust_llvm`, which is used in three places:
1. https://github.com/rust-lang/rust/blob/c5f92ce2a075c0f656362513ebbaf19006eab047/src/bootstrap/test.rs#L1553-L1554
2. The new assert I just added. Let's ignore that for now.
3. https://github.com/rust-lang/rust/blob/b3bc6bf31265ac10946a0832092dbcedf9b26805/src/bootstrap/compile.rs#L805-L806

3. defines LLVM_RUSTLLVM, which is eventually used in https://github.com/rust-lang/rust/blob/693c6311f06bf49ff0fd7b972d769634e4b330b1/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp#L308-L311; that bit looks correct. But 1. looks absolutely wrong even before this PR: it will pass `--system-llvm` for an in-tree commit with `submodules = false` and `patches = false`. How is `llvm_has_rust_patches` tested? how is it supposed to work? I ran `llvm-config --ldflags --system-llvm` locally and it didn't even exit successfully ...

jyn514 commented:
ugh I misread, it's a compiletest flag, not an llvm-config flag. so I think what was happening before is that some tests were silently being ignored, which we don't have any sort of checks for.

I'll fix 1 at the same time as I address your and pietro's comments.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3544409797/jobs/5951657015) [(plain)](https://github.com/rust-lang/rust/commit/7945458588a3f9c2bf048ac529be3f94a2afb7db/checks/9698251685/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/bootstrap/config.rs at line 1229:
                     target.llvm_config = Some(config.src.join(s));
                 }
                 if let Some(patches) = cfg.llvm_has_rust_patches {
-                    assert!(config.submodules.is_some(), "cannot set `llvm-hash-rust-patches` for a managed submodule");
+                        config.submodules.is_some(),
+                        config.submodules.is_some(),
+                        "cannot set `llvm-hash-rust-patches` for a managed submodule"
                     target.llvm_has_rust_patches = Some(patches);
                 }
                 }
                 if let Some(ref s) = cfg.llvm_filecheck {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/bootstrap/builder.rs" "/checkout/compiler/rustc_mir_build/src/thir/pattern/check_match.rs" "/checkout/compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs" "/checkout/compiler/rustc_mir_build/src/thir/constant.rs" "/checkout/src/bootstrap/config.rs" "/checkout/compiler/rustc_mir_build/src/lib.rs" "/checkout/compiler/rustc_mir_build/src/lints.rs" "/checkout/compiler/rustc_mir_build/src/thir/pattern/usefulness.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

jyn514 commented:
Ok, pushed a few changes. @pietroalbini @tmandry let me know if you think this will work better.

pietroalbini commented on config.toml.example:
```diff
@@ -292,6 +292,10 @@ changelog-seen = 2
 # on this runtime, such as `-C profile-generate` or `-C instrument-coverage`).
 #profiler = false
 
+# Use the optimized LLVM C intrinsics for `compiler_builtins`, rather than Rust intrinsics.
+# Requires the LLVM submodule to be managed by bootstrap (i.e. not external).
+#optimized-compiler-builtins = false
```
We should default to `true` IMO, and have people set to `false` if they have an unsupported configuration (and thus get an error message) or if they need to disable it for other reasons.

Setting this to `false` generates a broken compiler when using GCC < 9.4 as the linker on `aarch64-unknown-linux-gnu`, and there might be other targets that produce broken compilers with this flag off. I wouldn't want that to be the default, especially because it's super hard to debug those failures.

There should also be a comment explaining the downsides of setting it to `false`, and explaining what to do if LLVM is not managed by `bootstrap` to set it to `true` (see my other comment).

jyn514 replied:
If there were an easy way to tell when this breaks, I would have done that instead of adding the new configuration option. The original problem in https://github.com/rust-lang/rust/issues/101172 was that the cc crate doesn't support cross compiling to Windows; there may be other cases where it's broken I don't know about, and it's very hard to tell that it's related to this option.

Given that I would prefer to keep this defaulting to false. I'm happy to add the documentation you suggested.

pietroalbini replied:
My concern is that there is no good default for this flag. Setting it to `true` could cause [cross-compilation errors](https://github.com/rust-lang/rust/issues/101172) or [wrong architecture code to be included when cross-compiling](https://github.com/rust-lang/rust/issues/105065), while setting it to `false` results in the compiler being unable to produce any binary on a Tier 1 platform.

The problems when the setting is configured to `false` are IMO equally hard to diagnose as the problems when the setting is configured to `true`, I was only managed to figure out the missing flag was the cause by doing a PR-by-PR compile+bisect.

pietroalbini commented on src/bootstrap/compile.rs:
```diff
@@ -311,8 +309,18 @@ pub fn std_cargo(builder: &Builder<'_>, target: TargetSelection, stage: u32, car
+        // Using system llvm is not supported.
+        builder.update_submodule(&Path::new("src").join("llvm-project"));
+        let compiler_builtins_root = builder.src.join("src/llvm-project/compiler-rt");
+        if !compiler_builtins_root.exists() || builder.is_system_llvm(target) {
+            panic!(
+                "needed LLVM sources available to build `compiler-rt`, but they weren't present; consider enabling `build.submodules = true`, disabling `optimized-compiler-builtins`, or unsetting `llvm-config`"
+            );
+        }
```
Since `optimized-compiler-builtins` is required to produce a working compiler on the platforms we support, there should be a way to set it with a custom/system LLVM. Maybe having a `llvm.compiler-rt-source` setting to specify the path to the custom compiler-rt source code would work?

jyn514 replied:
This was *already* unsupported, it was just unclear because everything was implicit. The former logic was to enable the option when `llvm-project` was checked out and disable it when it's not. I don't think we should try and expand that here.

pietroalbini replied:
The lack of an escape hatch prevents building the compiler on a Tier 1 platform though, while before it worked (maybe by accident). I'm happy to contribute a `llvm.compiler-rt-source` flag myself.

pietroalbini commented on src/ci/docker/host-x86_64/x86_64-gnu-llvm-13/Dockerfile:
```diff
@@ -40,6 +40,7 @@ RUN sh /scripts/sccache.sh
 # We are disabling CI LLVM since this builder is intentionally using a host
 # LLVM, rather than the typical src/llvm-project LLVM.
 ENV NO_DOWNLOAD_CI_LLVM 1
+ENV EXTERNAL_LLVM 1
```
Note that the source could be downloaded with `apt-get source llvm-13`, so we should be able to provide the current source for this builder.

jyn514 replied:
I don't think that's a good idea, see https://github.com/rust-lang/rust/pull/102579#discussion_r1032930992.

tmandry commented:
@rustbot author (I think there are some outstanding concerns from Pietro to be addressed)

bors commented:
:umbrella: The latest upstream changes (presumably #108056) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

pietroalbini commented:
I don't have any more concerns if https://github.com/rust-lang/rust/issues/109064 is implemented (i.e. outline-atomics is removed from the aarch64 target spec)

jyn514 commented:
@pietroalbini how can I reproduce the error that you ran into? I'm using the old version of compiler-builtins without the new intrinsics (0.1.93) and I see in the paste that you were cross-compiling stage2 remote-test-server from x86_64-unknown-linux-gnu to aarch64-unknown-linux-gnu. But I can't replicate the error with `x b --stage 2 remote-test-server --target aarch64-unknown-linux-gnu` on either an x86 or aarch64 host.

jyn514 commented:
ah, you mentioned in https://github.com/rust-lang/rust/pull/101833#issuecomment-1270191721 that this needs gcc < 9.4 to replicate.

jyn514 commented:
ok, this is ready for review :) @tmandry @pietroalbini

jyn514 commented:
@bors rollup=iffy

emilyalbini reviewed (approved):
My concerns don't apply anymore now that outline-atomics are implemented in compiler_builtins. Haven't reviewed the code though, approving to remove the request for changes.

jyn514 commented:
blocked on https://github.com/rust-lang/rust/pull/113269 but i'd still appreciate a review from @tmandry on whether this handles `llvm.has_rust_patches` correctly

bors commented:
:umbrella: The latest upstream changes (presumably #112779) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tmandry commented on src/bootstrap/lib.rs:
```diff
@@ -863,18 +863,34 @@ impl Build {
-                (self.config.llvm_from_ci && target == self.config.build) || llvm_config.is_none()
+            // We're using pre-built LLVM and the user hasn't promised the patches match.
+            // This only has our patches if it's our managed, CI-built LLVM.
+            Some(Target { llvm_config: Some(_), .. }) => {
```
Actually, maybe you could use !is_system_llvm for these cases instead of repeating the matching logic?

tmandry commented on src/bootstrap/lib.rs:
```diff
@@ -863,18 +863,34 @@ impl Build {
+    fn is_system_llvm(&self, target: TargetSelection) -> bool {
+        match self.config.target_config.get(&target) {
+            Some(Target { llvm_config: Some(_), .. }) => {
+                !self.config.llvm_from_ci || target != self.config.build
```
nit: `!(a && b)` feels less confusing to me but ymmv

tmandry commented on src/bootstrap/lib.rs:
```diff
@@ -863,18 +863,34 @@ impl Build {
+                self.config.llvm_from_ci && target == self.config.build
             }
-            None => true,
+            Some(Target { llvm_config: None, llvm_has_rust_patches: None, .. }) | None => true,
```
nit: Seems less confusing to list these cases separately

tmandry reviewed (approved):
Looks good, thanks for checking! I left some nits, feel free to fix or ignore and r=me

bors commented:
:umbrella: The latest upstream changes (presumably #114148) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jyn514 commented:
https://github.com/rust-lang/team/pull/1051

jyn514 committed:
add and document a new `is_system_llvm` abstraction

jyn514 committed:
add a new `optimized_compiler_builtins` option

in particular, this makes the `c` feature for compiler-builtins an explicit opt-in, rather than silently detected by whether `llvm-project` is checked out on disk.
exposing this is necessary because the `cc` crate doesn't support cross-compiling to MSVC, and we want people to be able to run `x check --target foo` regardless of whether they have a c toolchain available.

this also uses the new option in CI, where we *do* want to optimize compiler_builtins.

the new option is off by default for the `dev` channel and on otherwise.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7321889025/job/19942790393) [(plain)](https://github.com/rust-lang/rust/commit/eb945e2e491d3f3a9ee09ef4fb3d3717e03d66dd/checks/19942790393/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_ae1c4366-4db1-4f50-9227-61124f63798a
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=fix-llvm-handling
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_ae1c4366-4db1-4f50-9227-61124f63798a
GITHUB_REF=refs/pull/102579/merge
GITHUB_REF_NAME=102579/merge
GITHUB_REF_PROTECTED=false
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Mon Dec 25 12:52:01 UTC 2023
  network time: Mon, 25 Dec 2023 12:52:01 GMT
  network time: Mon, 25 Dec 2023 12:52:01 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-missing-tools', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: rust.codegen-units-std := 1
---
##[endgroup]
WARNING: The `change-id` is missing in the `config.toml`. This means that you will not be able to track the major changes made to the bootstrap configurations.
NOTE: to silence this warning, add `change-id = 119124` at the top of `config.toml`
thread 'main' panicked at src/core/build_steps/compile.rs:407:13:
need LLVM sources available to build `compiler-rt`, but they weren't present; consider enabling `build.submodules = true`, disabling `optimized-compiler-builtins`, or unsetting `llvm-config`
Build completed unsuccessfully in 0:00:00
  local time: Mon Dec 25 12:52:36 UTC 2023
  network time: Mon, 25 Dec 2023 12:52:36 GMT
##[error]Process completed with exit code 1.
```

</details>

jyn514 commented:
i am going to remove the check that the in-tree llvm matches the llvm we're linking to, actually. i am not sure the difference matters, and it wasn't enforced before.

jyn514 commented:
r? bootstrap

jyn514 commented:
@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #119373) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

onur-ozkan reviewed (approved):
No concerns on my side. Currently, Jyn isn't actively working on the project. I will close this PR; open another one to cherry-pick the commits, resolve conflicts, and then r+ it.
